README.markdown
# AnalyticsGoo

AnalyticsGoo is a gem (or plugin) that provides server side non-javascript tracking using google analytics. 

## Install

gem install eyestreet-analytics_goo

## Example
   The simple usage is the following:
   
        tracker = AnalyticsGoo.config(:google_analytics, 
                                      :analytics_id => "UA-3536616-5", 
                                      :domain => "demo.eyestreet.com")

        tracker.track_page_view("/foo")

The first parameter passed to the config initializes the desired analytics adapter. Currently we only support
Google Analytics. The subsequent hash provide a name for the account, the analytics id, and a domain.

## Rails Usage

In your environment.rb add the following:

  config.gem "analytics_goo", :lib => "analytics_goo"
  

You can call things  in the same way as above, or you can mix in some additional rails specific functionality that is shown below.
In an intializer like config/initializers/analytics_goo.rb or in your appropriate environment.rb file

        AnalyticsGoo.config(:google_analytics, :analytics_id => "UA-3536616-5", 
                                               :domain => "demo.eyestreet.com", 
                                               :rails_core_mixins => true, 
                                               :environment => "production")

* analytics type - Currently we only support :google_analytics

The analytics option hash takes the following:

* :analytics_id
* :domain
* :rails_core_mixins - Defaults to false. If set to true then you get an accessor method on rails core classes for analytics_goo
* :environment - The RAILS_ENV environment that the analytics code should be called in. In all other environments it is a noop.

Then in your models, controllers, and mailers you can do the following:

    user = User.find(:first)
    user.analytics_goo.track_page_view("/found_my_first_user")

## Testing

Check test/test_helper.rb to make sure you have the required gems installed.

  rake test
  or
  autotest

# Credits

AnalyticsGoo is maintained by [Rob Christie](mailto:rob.christie@eyestreet.com) and is funded by [EyeStreet Software](http://www.eyestreet.com).


Copyright (c) 2009 rwc9u, Eyestreet Software released under the MIT license


CHANGELOG
[3b07757 | Tue Jun 02 19:39:50 UTC 2009] Rob Christie <robchristie@gmail.com>

 * More changes to support rails environments.

[3452913 | Tue Jun 02 18:39:51 UTC 2009] Rob Christie <robchristie@gmail.com>

 * Changed configuration to optionally allow mixins to Rails core classes. Changed to the use of markdown for my readme.

[e75017f | Tue Jun 02 07:18:37 UTC 2009] Rob Christie <robchristie@gmail.com>

 * Migrating to a gem installation.

[5b7698f | Tue Jun 02 06:00:49 UTC 2009] Rob Christie <robchristie@gmail.com>

 * Updated name of project.

[00a001f | Tue Jun 02 05:16:40 UTC 2009] Rob Christie <robchristie@gmail.com>

 * Updated based on naming and configuration changes.

[17529a1 | Mon Jun 01 20:10:51 UTC 2009] Rob Christie <robchristie@gmail.com>

 * Updated class so that random values are changed on each call to track_page_view.

[e0d96e7 | Mon Jun 01 19:14:54 UTC 2009] Rob Christie <robchristie@gmail.com>

 * Updated changelog.

[b6ec97c | Mon Jun 01 19:12:50 UTC 2009] Rob Christie <robchristie@gmail.com>

 * Renamed #track_it to #track_page_view, so you can eventually add other Analytics supported methods like #track_event.

[99e5798 | Mon Jun 01 18:55:32 UTC 2009] Rob Christie <robchristie@gmail.com>

 * Updated CHANGELOG after running script to pull and format messages from git.

[bbc664d | Mon Jun 01 18:48:16 UTC 2009] Rob Christie <robchristie@gmail.com>

 * Remove the AnalyticsConfig object and we now pass in a hash of initialization values. Also added support for passing in multiple hashes of initialization values.

[81ce0f8 | Wed May 20 14:29:06 UTC 2009] Rob Christie <robchristie@gmail.com>

 * initial commit



test/test_helper.rb
require 'rubygems'
require 'test/unit'
require 'active_support'
require 'active_support/test_case'
gem 'thoughtbot-shoulda', ">= 2.10.1"
require 'shoulda'
require 'mocha'


test/analytics_goo_test.rb
require 'test_helper'
require 'analytics_goo'

class AnalyticsGooTest < ActiveSupport::TestCase

  context "AnalyticsGoo " do
    setup do
      @analytics_config = { :analytics_id => "UA-2202604-2",:domain => "test.local" }
    end
    context "contains a GoogleAnalyticsAdapter which when passed initialization data" do
      should "be a valid class" do
        assert AnalyticsGoo::GoogleAnalyticsAdapter.new(@analytics_config)
      end
    end
    context "contains a GoogleAnalyticsAdapter when instantiated with the config method" do
      should "be a valid class" do
        assert AnalyticsGoo.config(:google_analytics, @analytics_config)
      end
    end
    should "raise an exception when instantiated with an invalid adapter" do
      assert_raise AnalyticsGoo::AnalyticsAdapterNotFound do
        AnalyticsGoo.config(:moogle_analytics, @analytics_config)
      end
    end
    context "can pass along an environment variable that" do
      setup do
        @analytics_config[:environment] = "test"
      end
     should "be passed along to the instantiated class" do
        assert_equal "test", AnalyticsGoo.config(:google_analytics, @analytics_config).env
      end
    end
  end

  context "AnalyticsGoo " do
    setup do
      @analytics_config = { :analytics_id => "UA-2202604-2",:domain => "test.local"}
      @ga = AnalyticsGoo::GoogleAnalyticsAdapter.new(@analytics_config)
      @ga.utmdt = "This is the page title"
      @ga.utmfl = "9.0 r124"
    end
    context "when initialized with an analytics id" do
      should "return that id" do
        assert_equal "UA-2202604-2", @ga.utmac
      end
    end
    context "when initialized with a domain" do
      should "return that domain" do
        assert_equal "test.local", @ga.domain
      end
    end
    context "that has been initialized" do
      should "have an urchin url" do
        assert_equal "http://www.google-analytics.com/__utm.gif", @ga.urchin_url
      end
    end
    context "when included and told that the url needs to be ssl" do
      should "have an https URL" do
        assert_equal "https://www.google-analytics.com/__utm.gif", @ga.urchin_url(true)
      end
    end
    context "creates an image URL that " do
      should "have a utmac equal to the specified analytics id" do
        assert_equal "UA-2202604-2", @ga.utmac
      end
      should "have a utmhn equal to the host name" do
        assert_equal "test.local", @ga.domain
      end
      should "have a utmfl equal to the flash version installed" do
        assert_equal "9.0 r124", @ga.utmfl
      end
      should "have a utmcs equal to the language encoding used for the browser." do
        assert_equal "UTF-8", @ga.utmcs
      end
      should "have a utmdt equal to the Page Title of the page" do
        assert_equal "This is the page title", @ga.utmdt
      end
      should "have a utmje equal to 1" do
        assert_equal "1", @ga.utmje
      end
      should "have a utmn equal to a random number that is generated to prevent caching" do
        assert @ga.utmn != @ga.utmn
      end
      should "have a utmsr equal to the screen resolution" do
        assert_equal "0x0", @ga.utmsr
      end
      should "have a utmsc equal to the screen color depth" do
        assert_equal "0-bit", @ga.utmsc
      end
      should "have a utmul equal to the browser language" do
        assert_equal "en-us", @ga.utmul
      end
      should "have a utmwv equal to the tracking code version" do
        assert_equal "4.3", @ga.utmwv
      end
      should "have a utmr equal to the referring page" do
        assert_equal "-", @ga.utmr
      end
      should "have a utmcr set to 1" do
        assert_equal "1", @ga.utmcr
      end
    end
  end
  context "Rails Analyics" do
    setup do
      @ga2 = AnalyticsGoo::GoogleAnalyticsAdapter.new(:analytics_id => "UA-3536616-5",:domain => "demo.mobilediscovery.com")
      @ga2.expects(:utmcc_cookie).times(2).returns(10000001)
      @ga2.expects(:utmcc_random).returns(1147483647)
      @now = Time.now.to_i
      @ga2.expects(:utmcc_time).times(4).returns(@now)
    end
    should "return a well formed utmcc cookie" do
      assert_equal "__utma%3D10000001.1147483647.#{@now}.#{@now}.#{@now}.10%3B%2B__utmz%3D10000001.#{@now}.1.1.utmcsr%3D(direct)%7Cutmccn%3D(direct)%7Cutmcmd%3D(none)%3B", @ga2.utmcc
    end
  end
  # TODO: mock this test. Currently it actually does do a request. Was using this to verify that it actually was working.
  context "An analytics tracking event" do
    setup do
      @ga3 = AnalyticsGoo::GoogleAnalyticsAdapter.new(:analytics_id => "UA-3536616-5",:domain => "demo.mobilediscovery.com")
    end
    context "makes a request for an image on the google analytics server" do
      setup do
        @resp = @ga3.track_page_view("/testFoo/myPage.html")
      end
      should "be a valid response" do
        assert @resp.is_a?(Net::HTTPOK)
      end
    end
  end
end


test/rails_analytics_test.rb
require 'test_helper'
require 'rails_analytics'

class RailsAnalyticsTest < ActiveSupport::TestCase

  context "RailsAnalytics contains a GoogleAnalyticsAdapter which when passed initialization data" do
    setup do
      @analytics_config = { :name => "test",:analytics_id => "UA-2202604-2",:domain => "test.local", :type => :google_analytics }
    end
    should "be a valid class" do
      assert RailsAnalytics::GoogleAnalyticsAdapter.new(@analytics_config)
    end
  end
  context "RailsAnalytics can be initialized with two analytics accounts" do
    setup do
      @analytics_config = {"test" => { :name => "test",:analytics_id => "UA-2202604-2",:domain => "test.local", :type => :google_analytics },
                           "test2" => { :name => "test2",:analytics_id => "UA-2202604-3",:domain => "test.local", :type => :google_analytics }
                          }
    end
    should "be a valid class" do
      assert RailsAnalytics::GoogleAnalyticsAdapter.new(@analytics_config)
    end
  end
  context "Rails Analytics " do
    setup do
      @analytics_config = { :name => "test",:analytics_id => "UA-2202604-2",:domain => "test.local", :type => :google_analytics }
      @ga = RailsAnalytics::GoogleAnalyticsAdapter.new(@analytics_config)
    end
    context "when initialized with an analytics id" do
      should "return that id" do
        assert_equal "UA-2202604-2", @ga.utmac("test")
      end
    end
    context "when initialized with a domain" do
      should "return that domain" do
        assert_equal "test.local", @ga.domain("test")
      end
    end
    context "that has been initialized" do
      should "have an urchin url" do
        assert_equal "http://www.google-analytics.com/__utm.gif", @ga.urchin_url
      end
    end
    context "when included and told that the url needs to be ssl" do
      should "have an https URL" do
        assert_equal "https://www.google-analytics.com/__utm.gif", @ga.urchin_url(true)
      end
    end
    context "creates an image URL that " do
      should "have a utmac equal to the specified analytics id" do
        assert_equal "UA-2202604-2", @ga.utmac("test")
      end
      should "have a utmhn equal to the host name" do
        assert_equal "test.local", @ga.domain("test")
      end
      should "have a utmfl equal to the flash version installed" do
        assert_equal "9.0 r124", @ga.utmfl
      end
      should "have a utmcs equal to the language encoding used for the browser." do
        assert_equal "UTF-8", @ga.utmcs
      end
      should "have a utmdt equal to the Page Title of the page" do
        assert_equal "This is the page title", @ga.utmdt
      end
      should "have a utmje equal to 1" do
        assert_equal "1", @ga.utmje
      end
      should "have a utmn equal to a random number that is generated to prevent caching" do
        assert @ga.utmn != @ga.utmn
      end
      should "have a utmsr equal to the screen resolution" do
        assert_equal "2400x1920", @ga.utmsr
      end
      should "have a utmsc equal to the screen color depth" do
        assert_equal "24-bit", @ga.utmsc
      end
      should "have a utmul equal to the browser language" do
        assert_equal "en-us", @ga.utmul
      end
      should "have a utmwv equal to the tracking code version" do
        assert_equal "4.3", @ga.utmwv
      end
      should "have a utmr equal to the referring page" do
        assert_equal "-", @ga.utmr
      end
      should "have a utmcr set to 1" do
        assert_equal "1", @ga.utmcr
      end
    end
  end
  context "Rails Analyics" do
    setup do
      @ga2 = RailsAnalytics::GoogleAnalyticsAdapter.new(:name => :test, :analytics_id => "UA-3536616-5",:domain => "demo.mobilediscovery.com",:type => :google_analytics)
      @ga2.expects(:utmcc_cookie).times(2).returns(10000001)
      @ga2.expects(:utmcc_random).returns(1147483647)
      @now = Time.now.to_i
      @ga2.expects(:utmcc_time).times(4).returns(@now)
    end
    should "return a well formed utmcc cookie" do
      assert_equal "__utma%3D10000001.1147483647.#{@now}.#{@now}.#{@now}.10%3B%2B__utmz%3D10000001.#{@now}.1.1.utmcsr%3D(direct)%7Cutmccn%3D(direct)%7Cutmcmd%3D(none)%3B", @ga2.utmcc
    end
  end
  # TODO: mock this test. Currently it actually does do a request. Was using this to verify that it actually was working.
  context "An analytics tracking event" do
    setup do
      @ga3 = RailsAnalytics::GoogleAnalyticsAdapter.new(:name => :test, :analytics_id => "UA-3536616-5",:domain => "demo.mobilediscovery.com",:type => :google_analytics)
    end
    context "makes a request for an image on the google analytics server" do
      setup do
        @resp = @ga3.track_page_view("/testFoo/myPage.html")
      end
      should "be a valid response" do
        assert @resp[0].is_a?(Net::HTTPOK)
      end
    end
  end
end


lib/analytics_goo.rb
require 'uri'
require 'cgi'
require 'net/http'
require 'activesupport'
require 'analytics_goo/google_analytics_adapter'

module AnalyticsGoo
  # generic adapter not found exception
  class AnalyticsAdapterNotFound < StandardError
  end


  # Factory for returning the appropriate analytics object. The <tt>type</tt> is
  # a symbol that defines the type of analytics tracker you want to create. Currently,
  # the only acceptable value is :google_analytics. The <tt>analytics</tt> hash holds
  # the required keys of analytics_id and the domain of the adapter that you are configuring.
  # Setting the key <tt>rails_core_mixins</tt> to
  # true will mixin an accessor for this object into the core rails framework classes (active_record, action_controller and action_mailer). The default
  # for this is false so that there are no rails dependencies on the gem. The key <tt>environment</tt> by default is set to "production". The
  # rails mixin functionality checks this value against its RAILS_ENV then it calls out to the analytics only if the value is nil or matches.
  def self.config(type, analytics = {})
    begin
      s = type.to_s + "_adapter"
      adapter = "AnalyticsGoo::" + s.camelize
      analytics[:environment] = "production" if analytics[:environment].nil?
      if analytics[:rails_core_mixins] == true
        analytics[:noop] = (RAILS_ENV != analytics[:environment])
        tracker = adapter.constantize.new(analytics)
        for framework in ([ :active_record, :action_controller, :action_mailer ])
          framework.to_s.camelize.constantize.const_get("Base").send :include, AnalyticsGoo::InstanceMethods
        end
        silence_warnings { Object.const_set "ANALYTICS_TRACKER", tracker }
      else
        tracker = adapter.constantize.new(analytics)
      end
      tracker
    rescue StandardError
      raise AnalyticsAdapterNotFound
    end
  end

  module InstanceMethods
    # any methods here will apply to instances
    def analytics_goo
      if defined?(ANALYTICS_TRACKER)
          ANALYTICS_TRACKER
      else
        nil
      end
    end
  end
end



lib/rails_analytics.rb
require 'uri'
require 'cgi'
require 'net/http'
require 'rails_analytics/google_analytics_adapter'

module RailsAnalytics
  # generic adapter not found exception
  class AnalyticsAdapterNotFound < StandardError
  end

  def self.config(analytics)
    begin
      s = analytics[:type].to_s + "_adapter"
      adapter = "RailsAnalytics::" + s.camelize
      tracker = adapter.constantize.new(analytics)
      silence_warnings { Object.const_set "ANALYTICS_TRACKER", tracker }
    rescue StandardError
      raise AnalyticsAdapterNotFound
    end
  end

  module InstanceMethods
    # any methods here will apply to instances
    def track_page_view(path, name=nil)
      if defined?(ANALYTICS_TRACKER)
        ANALYTICS_TRACKER.track_page_view(path,name)
      else
        nil
      end
    end
  end
end



lib/analytics_goo/google_analytics_adapter.rb
# AnalyticsGoo
# The following came from :
#  http://code.google.com/apis/analytics/docs/tracking/gaTrackingTroubleshooting.html

# The following parameters are currently not implemented but do not affect basic usage.
# utme  X10 Data Parameter  Value is encoded.

# utmipc
#   Product Code. This is the sku code for a given product.
#   utmipc=989898ajssi
# utmipn
#   Product Name, which is a URL-encoded string.  utmipn=tee%20shirt
# utmipr
#   Unit Price. Set at the item level. Value is set to numbers only in U.S. currency format.
#   utmipr=17100.32
# utmiqt
#   Quantity.   utmiqt=4
# utmiva
#   Variations on an item. For example: large, medium, small, pink, white, black, green. String is URL-encoded.
#   utmiva=red;
# utmt
#   A special type variable applied to events, transactions, items and user-defined variables.  utmt=Dog%20Owner
# utmtci
#   Billing City  utmtci=San%20Diego
# utmtco
#   Billing Country   utmtco=United%20Kingdom
# utmtid
#   Order ID, URL-encoded string.   utmtid=a2343898
# utmtrg
#   Billing region, URL-encoded string.   utmtrg=New%20Brunswick
# utmtsp
#   Shipping cost. Values as for unit and price.  utmtsp=23.95
# utmtst
#   Affiliation. Typically used for brick and mortar applications in ecommerce.   utmtst=google%20mtv%20store
# utmtto
#   Total. Values as for unit and price.  utmtto=334.56
# utmttx
#   Tax. Values as for unit and price.  utmttx=29.16

module AnalyticsGoo
  class GoogleAnalyticsAdapter
    attr_accessor :domain, :analytics_id, :env, :noop, :utmdt, :utmfl
    # utmdt
    #   Page title, which is a URL-encoded string.  utmdt=analytics%20page%20test
    # utmfl
    #   Flash version

    def initialize(ac)
      # sets the environment that this should be run in
      @analytics_id = ac[:analytics_id]
      @domain = ac[:domain]
      @env = ac[:environment]
      @noop = ac[:noop] || false
      @utmdt = ac[:utmdt] || ""
      @utmfl = ""
    end

    GA_DOMAIN = "www.google-analytics.com"
    #    GA_DOMAIN = "dev.www.mobilediscovery.com"
    GA_IMAGE = "__utm.gif"

    def urchin_url(ssl = false)
      protocol = (ssl == true) ? "https":"http"
      "#{protocol}://#{GA_DOMAIN}/#{GA_IMAGE}"
    end

    # utmac   Account String. Appears on all requests.    utmac=UA-2202604-2
    def utmac
      self.analytics_id
    end

    # utmhn
    #   Host Name, which is a URL-encoded string.   utmhn=x343.gmodules.com
    def utmhn
      self.domain
    end

    # utmcs
    #   Language encoding for the browser. Some browsers don't set this, in which case it is set to "-"
    #   utmcs=ISO-8859-1
    def utmcs
      "UTF-8"
    end

    # utmje
    #   Indicates if browser is Java-enabled. 1 is true.  utmje=1
    def utmje
      "1"
    end

    # utmn
    #   Unique ID generated for each GIF request to prevent caching of the GIF image.   utmn=1142651215
    def utmn
      ActiveSupport::SecureRandom.random_number(9999999999)
    end

    # utmsc
    #   Screen color depth  utmsc=24-bit
    def utmsc
      "0-bit"
    end

    # utmsr
    #   Screen resolution   utmsr=2400x1920&
    def utmsr
      "0x0"
    end

    # utmul
    #   Browser language.   utmul=pt-br
    def utmul
      "en-us"
    end

    # utmwv
    #   Tracking code version   utmwv=1
    def utmwv
      "4.3"
    end

    # utmp
    #   Page request of the current page.   utmp=/testDirectory/myPage.html
    #     def utmp
    #       self.path
    #     end

    # utmr
    #   Referral, complete URL.   utmr=http://www.example.com/aboutUs/index.php?var=selected
    def utmr
      "-"
    end


    # may not be needed

    # utmcn   Starts a new campaign session. Either utmcn or utmcr is present on any given request. Changes the campaign tracking data; but does not start a new session
    #   utmcn=1
    def utmcn
      "1"
    end

    # utmcr
    #   Indicates a repeat campaign visit. This is set when any subsequent clicks occur on the same link. Either utmcn or utmcr is present on any given request.
    #   utmcr=1
    def utmcr
      "1"
    end

    # utmcc
    #   Cookie values. This request parameter sends all the cookies requested from the page.
    #   utmcc=__utma%3D117243.1695285.22%3B%2B __utmz%3D117945243.1202416366.21.10. utmcsr%3Db%7C utmccn%3D(referral)%7C utmcmd%3Dreferral%7C utmcct%3D%252Fissue%3B%2B
    def utmcc
      "__utma%3D#{utmcc_cookie}.#{utmcc_random}.#{utmcc_time}.#{utmcc_time}.#{utmcc_time}.10%3B%2B__utmz%3D#{utmcc_cookie}.#{utmcc_time}.1.1.utmcsr%3D(direct)%7Cutmccn%3D(direct)%7Cutmcmd%3D(none)%3B"
    end

    # send a request to get the image from google
    def track_page_view(path)
      res = ""
      unless @noop == true
        res = track_it(path)
      end
      res
    end

    protected
    def track_it(path)
#       puts "/__utm.gif?utmwv=#{self.utmwv}&utmn=#{self.utmn}&utmhn=#{self.utmhn}&utmcs=#{self.utmcs}&utmsr=#{self.utmsr}&utmsc=#{self.utmsc}&utmul=#{self.utmul}&utmje=#{self.utmje}&utmfl=#{URI::escape(self.utmfl)}&utmdt=#{URI::escape(self.utmdt)}&utmhid=#{utmhid}&utmr=#{self.utmr}&utmp=#{path}&utmac=#{self.utmac}&utmcc=#{self.utmcc} \n"
      Net::HTTP.get_response(GA_DOMAIN,"/__utm.gif?utmwv=#{self.utmwv}&utmn=#{self.utmn}&utmhn=#{self.utmhn}&utmcs=#{self.utmcs}&utmsr=#{self.utmsr}&utmsc=#{self.utmsc}&utmul=#{self.utmul}&utmje=#{self.utmje}&utmfl=#{URI::escape(self.utmfl)}&utmdt=#{URI::escape(self.utmdt)}&utmhid=#{utmhid}&utmr=#{self.utmr}&utmp=#{path}&utmac=#{self.utmac}&utmcc=#{self.utmcc}")
    end

    def utmcc_cookie
      ActiveSupport::SecureRandom.random_number(89999999) + 10000000
    end

    def utmcc_random
      ActiveSupport::SecureRandom.random_number(1147483647) + 1000000000
    end

    def utmcc_time
      Time.new.to_i
    end

    def utmhid
      ActiveSupport::SecureRandom.random_number(999999999)
    end
  end
end
